include/proxy-wasm/context_interface.h

```diff
@@ -0,0 +1,633 @@
+   * incoming port that was accepted and for which the proxy is acting as a server).
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onDownstreamData() = 0;
+
+  /**
+   * Call on a stream Context to indicate that data has arrived from upstream (e.g. on the
+   * outgoing port that the proxy connected and for which the proxy is acting as a client).
+   * @param data_length the amount of data in the buffer.
+   * @param end_of_stream is true if no more data will be arriving.
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onUpstreamData() = 0;
+
+  /*
+   * The source of the close event.
+   * Unknown is when the source is not known.
+   * Local is when the close was initiated by the proxy.
+   * Remote is when the close was received from the peer.
+   */
+  enum class CloseType : uint32_t {
+    Unknown = 0,
+    Local = 1,
+    Remote = 2,
+  };
+
+  /**
+   * Call on a stream context to indicate that the downstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onDownstreamConnectionClose(CloseType close_type) = 0;
+
+  /**
+   * Call on a stream context to indicate that the upstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onUpstreamConnectionClose(CloseType close_type) = 0;
+
+  // Call when the stream closes. See RootInterface.
+  virtual bool onDone() = 0;
+
+  // Call when the stream status has finalized, e.g. for logging. See RootInterface.
+  virtual void onFinalized() = 0;
+
+  // Call just before the Context is deleted. See RootInterface.
+  virtual void onDelete() = 0;
+};
+
+/**
+ * General Stream interface (e.g. HTTP, Network).
+ **/
+struct StreamInterface {
+
+  // Continue processing a request e.g. after returning ProxyAction::Pause.
+  virtual WasmResult continueStream(WasmStreamType type) = 0;
+
+  /**
+   * Provides a BufferInterface to be used to return buffered data to the VM.
+   * @param type is the type of buffer to provide.
+   */
+  virtual const BufferInterface *getBuffer(WasmBufferType type) = 0;
+
+  /**
+   * Provides the end-of-stream status of the given flow (if any) or false.  End of stream occurs
+   * when a connection is closed (or half-closed) either locally or remotely.
+   * @param stream_type is the flow
+   */
+  virtual bool end_of_stream(WasmStreamType type) = 0;
+};
+
+// Header/Trailer/Metadata Maps
+struct HeaderInterface {
+  /**
+   * Add a key-value pair to a header map.
+   * @param type of the header map.
+   * @param key is the key (header).
+   * @param value is the value (header value).
+   */
+  virtual WasmResult addHeaderMapValue(WasmHeaderMapType type, string_view key,
+                                       string_view value) = 0;
+
+  /**
+   * Get a value from to a header map.
```

Comment:
What is the expected performance? O(1) or O(n)? It's important to have callers aware of this IMHO.